### PR TITLE
Reject identifiers with special characters.

### LINKED ISSUE #260: Disallow identifier names containing `@`

### PR DESCRIPTION
**Pull request by bearbin**
Fixes #260

### CODE REVIEW

File: Sources/Parser/Tokenizer.swift [hunk 1]
```diff
@@ -62,6 +62,10 @@ public struct Tokenizer {
       } else if component.first == "@" {
         // The token is a function attribute.
         tokens.append(Token(kind: .attribute(String(component.dropFirst())), sourceLocation: sourceLocation))
+      } else if let _ = component.rangeOfCharacter(from: invalidIdentifierChars) {
+        // The token is not valid.
+        print(component)
```
Comment (franklinsch) [inline Sources/Parser/Tokenizer.swift]: Is this print meant to be here?

File: Sources/Parser/Tokenizer.swift [hunk 2]
```diff
@@ -62,6 +62,10 @@ public struct Tokenizer {
...
+        fatalError("Parse error: invalid token!")
```
Comment (franklinsch) [inline Sources/Parser/Tokenizer.swift]: No need for the `!`

File: Sources/Parser/Tokenizer.swift [hunk 3]
```diff
@@ -210,6 +214,12 @@ public struct Tokenizer {
     return CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "$"))
   }
 
+  var invalidIdentifierChars: CharacterSet {
+    var ichars = identifierChars
+    ichars.invert()
+    return ichars
+  }
```
Comment (franklinsch) [inline Sources/Parser/Tokenizer.swift]: https://developer.apple.com/documentation/foundation/nscharacterset/1414025-inverted

File: Sources/Parser/Tokenizer.swift (hunk 1 above)
Comment (bearbin) [reply Sources/Parser/Tokenizer.swift]: No, will fix.

**Comment by bearbin** [comment]
I wasn't sure if rejecting the identifiers in the parser was the right move, but I've fixed the issues you raised.

**Comment by bearbin** [comment]
The reason the check for $ is done in the semanticAnalyser is because $ is allowed in identifiers for the standard library, but not elsewhere. Other special characters aren't allowed anywhere, so they wouldn't be valid identifiers in any case. So I thought the parser should reject things that are always going to be invalid. 

I can move the check into the SemAn, and get better errors too, just confirm that's what you want me to do.

**Comment by franklinsch** [comment]
Yes, I think moving to the Semantic Analyser makes more sense; I think it's a semantic issue rather than a syntax issue. It'd also give us better error messages.

**Comment by bearbin** [comment]
This is ready to review again.

**Comment by franklinsch** [comment]
I just noticed `_` is considered to be a special character. Why is that? Using underscores in identifier names is accepted in almost all programming languages.

**Comment by bearbin** [comment]
It must have crept in by mistake, I'll remove it.

**Comment by bearbin** [comment]
Resolved the issues.

**Pull request merged**